stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+                                if ioe1.kind() == io::ErrorKind::UnexpectedEof && ioe2.kind() == io::ErrorKind::UnexpectedEof {
+                                    // out of bytes
+                                    Err(CodecError::UnderflowError("Not enough bytes to form a HTTP request or response".to_string()))
+                                }
+                                else {
+                                    Err(CodecError::DeserializeError(format!("Neither a HTTP request ({:?}) or HTTP response ({:?})", ioe1, ioe2)))
+                                }
+                            },
+                            (e_req, e_res) => Err(CodecError::DeserializeError(format!("Failed to decode HTTP request or HTTP response (request error: {:?}; response error: {:?})", &e_req, &e_res)))
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl MessageSequence for StacksHttpMessage {
+    fn request_id(&self) -> u32 {
+        // there is at most one in-flight HTTP request, as far as a Connection<P> is concerned
+        HTTP_REQUEST_ID_RESERVED
+    }
+
+    fn get_message_name(&self) -> &'static str {
+        "StachsHttpMessage"
+    }
+}
+
+/// A partially-decoded, streamed HTTP message (response) being received.
+/// Internally used by StacksHttp to keep track of chunk-decoding state.
+#[derive(Debug, Clone, PartialEq)]
+struct StacksHttpRecvStream {
+    state: HttpChunkedTransferReaderState,
+    data: Vec<u8>,
+    total_consumed: usize, // number of *encoded* bytes consumed
+}
+
+impl StacksHttpRecvStream {
+    pub fn new(max_size: u64) -> StacksHttpRecvStream {
+        StacksHttpRecvStream {
+            state: HttpChunkedTransferReaderState::new(max_size),
+            data: vec![],
+            total_consumed: 0,
+        }
+    }
+
+    /// Feed data into our chunked transfer reader state.  If we finish reading a stream, return
+    /// the decoded bytes (as Some(Vec<u8>) and the total number of encoded bytes consumed).
+    /// Always returns the number of bytes consumed.
+    pub fn consume_data<R: Read>(
+        &mut self,
+        fd: &mut R,
+    ) -> Result<(Option<(Vec<u8>, usize)>, usize), NetError> {
+        let mut consumed = 0;
+        let mut blocked = false;
+        while !blocked {
+            let mut decoded_buf = vec![0u8; CHUNK_BUF_LEN];
+            let (read_pass, consumed_pass) = match self.state.do_read(fd, &mut decoded_buf) {
+                Ok((0, num_consumed)) => {
+                    trace!(
+                        "consume_data blocked on 0 decoded bytes ({} consumed)",
+                        num_consumed
+                    );
+                    blocked = true;
+                    (0, num_consumed)
+                }
+                Ok((num_read, num_consumed)) => (num_read, num_consumed),
+                Err(e) => match e.kind() {
```

Comment:
I think instead of doing a second match you can do 
```rust
    Ok((0, num_consumed)) => {..},
    Err(io::ErrorKind::WouldBlock) | Err(io::ErrorKind::TimedOut) => {
        trace!("consume_data blocked on read error");
        blocked = true;
        (0, 0)
    }
    _ => {
        return Err(NetError::ReadError(e));
    }
}
```

Reply:
Fixed